### PR TITLE
Fix uniform on CUDA tensor to return in range [0, 1)

### PR DESCRIPTION
**Pull request by colesbury**
The curand_uniform function returns the range (0, 1]. Most RNG APIs have
the opposite bounds. Fixup the values in uniform_() so that they fall in
the more common bounds.

See #3424 and #2740

### CODE REVIEW

**Comment by colesbury** [comment]
cc @ngimel

**Comment by colesbury** [comment]
```python
import torch
x = torch.cuda.FloatTensor(int(100*1e6));
print(x.uniform_().max())
x = torch.cuda.HalfTensor(int(100*1e6));
print(x.uniform_().max())
```

```
0.9999999403953552
0.99951171875
```

File: aten/src/THC/THCTensorRandom.cu [hunk 1]
```diff
@@ -143,7 +154,7 @@ GENERATE_KERNEL2(generate_cauchy, float, double median, double sigma, float, cur
 GENERATE_KERNEL2(generate_cauchy, double, double median, double sigma, double, curand_uniform_double, (double)(median + sigma * tan(M_PI*(x-0.5))))
 
 #ifdef CUDA_HALF_TENSOR
-GENERATE_KERNEL2(generate_uniform, half, double a, double b, float, curand_uniform, (ScalarConvert<float, half>::to(x * (b-a) + a)))
+GENERATE_KERNEL2(generate_uniform, half, double a, double b, float, curand_uniform, (__float2half_rd(reverse_bounds(x) * (b-a) + a)))
```
Comment (fmassa) [inline aten/src/THC/THCTensorRandom.cu]: Looks like this breaks for CUDA 7.5 (`__float2half_rd` is undefined).
Are we still supporting CUDA 7.5?

Comment (soumith) [reply aten/src/THC/THCTensorRandom.cu]: yea we need to fix it.

**Pull request merged**